File: matiasmasso/Mat.NET/DTO/DTOExtracte.cs
using MatHelperStd;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOExtracte
    {
        public DTOExercici Exercici { get; set; }
        public DTOPgcCta Cta { get; set; }
        public DTOContact Contact { get; set; }
        public List<DTOCcb> Ccbs { get; set; }
        public List<int> Years { get; set; }

        public static DTOExtracte Factory(DTOExercici oExercici, DTOPgcCta oCta, DTOContact oContact)
        {
            DTOExtracte retval = new DTOExtracte();
            {
                var withBlock = retval;
                withBlock.Exercici = oExercici;
                withBlock.Cta = oCta;
                withBlock.Contact = oContact;
            }
            return retval;
        }

        public static string Filename(List<DTOCcb> oCcbs, DTOLang oLang)
        {
            string retval = "M+O.extracte.xlsx";
            if (oCcbs.Count > 0)
            {
                DTOCcb oCcb = oCcbs.First();
                DTOPgcCta oCta = oCcb.Cta;
                string sCta = DTOPgcCta.FullNom(oCta, oLang);
                DTOContact oContact = oCcb.Contact;
                if (oContact == null)
                    retval = string.Format("M+O.extracte.{0}.xlsx", sCta);
                else
                {
                    string sContact = TextHelper.CleanForUrl(oContact.FullNom);
                    if (sContact == "")
                        retval = string.Format("M+O.extracte.{0}.xlsx", sCta);
                    else
                        retval = string.Format("M+O.extracte.{0}.{1}.xlsx", sCta, sContact).Replace(TextHelper.VbChr(34), "");
                }
            }
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOExtracte.vb
Public Class DTOExtracte
    Property Exercici As DTOExercici
    Property Cta As DTOPgcCta
    Property Contact As DTOContact
    Property Ccbs As List(Of DTOCcb)
    Property Years As List(Of Integer)

    Shared Function Factory(oExercici As DTOExercici, oCta As DTOPgcCta, oContact As DTOContact)
        Dim retval As New DTOExtracte
        With retval
            .Exercici = oExercici
            .Cta = oCta
            .Contact = oContact
        End With
        Return retval
    End Function

    Shared Function Filename(oCcbs As List(Of DTOCcb), oLang As DTOLang) As String
        Dim retval As String = "M+O.extracte.xlsx"
        If oCcbs.Count > 0 Then
            Dim oCcb As DTOCcb = oCcbs.First
            Dim oCta As DTOPgcCta = oCcb.Cta
            Dim sCta As String = DTOPgcCta.FullNom(oCta, oLang)
            Dim oContact As DTOContact = oCcb.Contact
            If oContact Is Nothing Then
                retval = String.Format("M+O.extracte.{0}.xlsx", sCta)
            Else
                Dim sContact As String = TextHelper.CleanForUrl(oContact.FullNom)
                If sContact = "" Then
                    retval = String.Format("M+O.extracte.{0}.xlsx", sCta)
                Else
                    retval = String.Format("M+O.extracte.{0}.{1}.xlsx", sCta, sContact).Replace(TextHelper.VbChr(34), "")
                End If
            End If
        End If
        Return retval
    End Function

End Class
